### DISCUSSION TITLE
Useful general widgets

### DISCUSSION THREAD

**Discussion by apirogov** [Materials-Data-Science-and-Informatics/metador-core#49] [Ideas]
Here we collect a wish list of widgets for the UI, and if they become part of metador-core or rather their project.

**Widgets:**

**Basic:**
- [X] Video (panel)
- [X] Audio (panel)
- [X] PDF (panel)

Text-based:

- [X] Markdown (panel has [one](https://panel.holoviz.org/reference/panes/Markdown.html))
- [X] HTML (panel has [one](https://panel.holoviz.org/reference/panes/HTML.html))
- [X] JSON (panel has [one](https://panel.holoviz.org/reference/panes/JSON.html)) 
- [X] Code/Plaintext (panel Ace widget)
- [X] CSV (panel Tabulator widget)
- [ ] JSON widget to show node metadata (i.e. not the payload, but content of another schema) - how should it work?

more fancy:

- [ ] basic configurable 2D Plots (designing a good metadata schema for configuration might be a challenge)
- [ ] Static Jupyter Notebook rendering (like Github does, useful for working around non-existing widgets / custom static visualizations)
- [ ] Graph visualization with [sigma.js](https://www.sigmajs.org/)

**special:**
- [ ] file Download
      *We can get files out (API exists), but not folders and not arbitrary entities - its HDF5!*
      But a "download this file" button could be part of a generic embedded file widget, or part of the tree view component.
      *Download the whole Project Data is of course also possible.*

**Widget consumers:**
- [X] Standard dashboard for a Metador container (uses installed widgets and instantiates for nodes marked for dashboard view)
- [ ] Container tree-view, interactive display (this could also bis displayed as a graph)

**Advanced**
- [ ] VTK display (probably for relevant special cases) (panel also has something there [already](https://panel.holoviz.org/reference/panes/VTK.html) and VTKVolume, VTKJS.
- [ ] Atomic structure display (there should be something out there to be reusable)
- [ ] advanced interactive configurable plots and charts (echarts-based?)
- [ ] Allow people to provide a 'discover' section for their container, i.e their own custom dashboard using metador compatible widgets

**Comment by broeder-j** [comment]
Hi @apirogov  I added several widgets to the wish list above.

**Comment by apirogov** [comment]
I commented above on the dashboard and download thingy, the tree-style explorer view is something I also had on my mind, but I guess it could be quite some work and its not super high priority for now.

**Comment by broeder-j** [comment]
@FionaDmello Could you comment, or point to the metador VTK widget development and how this differs from what is already there in panel. 
Maybe it makes sense in some cases to create a new panel widgets, instead.

Also in general since fancy [ECharts](https://echarts.apache.org/en/index.html) can be embeded in Panel it might also make sense to use those instead. For them I will learn javascript ;-)
A subset is available from python directly [pyecharts](https://github.com/pyecharts/pyecharts/blob/master/README.en.md)
Since one could execute in notebooks mixed code, it would also work in the browser and notebook, but not from script/command line).

**Comment by FionaDmello** [comment]
Thanks @broeder-j for compiling this list of useful widgets! The metador VTK widget is yet to be defined within metador-core. So far, the development of this widget mainly focused on testing whether we could rely on [Panel](https://panel.holoviz.org/index.html) for visualization, to a reasonable extent, of scientific data stored in VTK files.
 
Currently, the custom widgets set up to test Panel are built on [Bokeh](https://panel.holoviz.org/reference/panes/Bokeh.html). So the question at this point would be - do we really want/need to depend on [ECharts](https://panel.holoviz.org/reference/panes/ECharts.html) rather than Bokeh? Why? Despite the existence of [pyecharts](https://github.com/pyecharts/pyecharts/blob/master/README.en.md), would the decision to build our visualization widgets with ECharts and pyecharts instead of Bokeh, be reasonable and acceptable to the scientific community we aim to serve? 

I am interested in finding out the difference and advantages of using ECharts and pyecharts over Bokeh. I will take some time to recreate a set of existing custom widgets with ECharts and see how it goes ;-)

**Reply by broeder-j** [reply]
As far as I understood we are depending mainly on Panel. So everything which works with panel is fine. I was not suggesting to just replace bokeh with echarts. I was just pointing out that one is free to write widgets in echarts. Also you can well assume that most widgets will be written by us, or will be reuses of other work which is out there. I have no strong opinion on this.

**Comment by broeder-j** [comment]
Maybe from these one could also recyle something: https://www.osscar.org/tutorial/tools.html#
Or this adapted: https://github.com/materialscloud-org/structure-property-visualizer

or this (structure and BZ visualizer) https://nomad-coe.github.io/materia/index.html
